### ISSUE TITLE
Do not access Python data structures from weave/cython

### ISSUE THREAD

**Issue by mstimberg** [brian-team/brian2#970]
Currently, a major slow down in run time mode is our access of Python data structures from our generated C++ code (weave/Cython). For example, monitors and synapse generation resize dynamical arrays, and spike propagation accesses the `SpikeQueue`. For the latter, we are already using a C implementation by default, but accessing it from C++ code goes via its Python wrapper, adding unnecessary overhead.

I propose that we:
* [ ] write a C++ implementation of the `DynamicArray` class and provide a Cython wrapper around it so that it can still be used from numpy
* [ ] remove the Python implementation of the `SpikeQueue` and always use the Cython version
* [ ] directly access the underlying C++ implementation of the `SpikeQueue` from C++ code

The only disadvantage of this that I am seeing is that you can no longer run Brian from source without compiling the `SpikeQueue` and `DynamicArray`. On the other hand, if you are using the code directly from github instead of from conda or pypi (where we provide binary packages for Windows and OS X already, and can potentially provide compiled packages for Linux as well), you are probably also able to call a command to compile the extensions manually.

I see two big advantages of doing things this way:
* Improved performance for weave/Cython (always good)
* we can use the same set of templates for the code objects in weave and C++ standalone

If, in the longer run, we also get rid of Cython by writing our own weave replacement, this would greatly reduce the number of templates and the code you have to write when implementing a new feature.

**Comment by thesamovar** [comment]
Very good idea. Strictly speaking, we don't need to get rid of our Python implementation: that could still be a fallback. Not sure if that's necessary any longer given how easy it is to install binaries now.

**Comment by mstimberg** [comment]
True, we could keep the Python implementation as a fallback for the `numpy` target. I'm also not sure whether this is worth it, my main argument against it would be that it means more code, more testing, and more possible bugs.

**Comment by thesamovar** [comment]
As a general rule I agree with that, although this particular piece of code is now very stable. I think we can be pretty confident about it?
